src/wp-includes/class-wp-query.php

```diff
@@ -2517,39 +2524,69 @@ public function get_posts() {
 				$where .= " AND ($where_status)";
 			}
 		} elseif ( ! $this->is_singular ) {
-			$where .= " AND ({$wpdb->posts}.post_status = 'publish'";
-
-			// Add public states.
-			$public_states = get_post_stati( array( 'public' => true ) );
-			foreach ( (array) $public_states as $state ) {
-				if ( 'publish' === $state ) { // Publish is hard-coded above.
-					continue;
-				}
-				$where .= " OR {$wpdb->posts}.post_status = '$state'";
+			if ( 'any' === $post_type ) {
+				$queried_post_types = get_post_types( array( 'exclude_from_search' => false ) );
+			} elseif ( is_array( $post_type ) ) {
+				$queried_post_types = $post_type;
```

Comment:
This should work and we can write a test for that. Just note that this is not currently supported, it would be yet another fix

Reply:
Ah, looks like I got post type and status bugs confused again. Ignore my comment and it can be left for another day so this doesn't become a three ticket PR :D